macaroons/store.go

```diff
@@ -118,6 +122,77 @@ func (r *RootKeyStorage) CreateUnlock(password *[]byte) error {
 	})
 }
 
+// ChangeEncKey updates the encryption key used to encrypt/decrypt the root keys
+// stored.
+func (r *RootKeyStorage) ChangeEncKey(old, new []byte) error {
+	return r.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(rootKeyBucketName)
+		dbKey := bucket.Get(encryptedKeyID)
+
+		// First, we'll make sure the macaroon store has been created.
+		if len(dbKey) == 0 {
+			return ErrStoreNotCreated
+		}
+
+		// If it is, we'll verify that it is correct and create a new
+		// encryption key with the new password.
+		oldEncKey := &snacl.SecretKey{}
+		err := oldEncKey.Unmarshal(dbKey)
+		if err != nil {
+			return err
+		}
+
+		err = oldEncKey.DeriveKey(&old)
+		if err != nil {
+			return err
+		}
+
+		newEncKey, err := snacl.NewSecretKey(
+			&new, snacl.DefaultN, snacl.DefaultR, snacl.DefaultP,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Then, for all the root keys in our database, we'll update
+		// their encrypted form values to use the new encryption key.
+		err = bucket.ForEach(func(k, v []byte) error {
+			// We'll skip updating the encryption key itself.
+			if bytes.Equal(k, encryptedKeyID) {
+				return nil
+			}
+
+			decrypted, err := r.encKey.Decrypt(v)
+			if err != nil {
+				return err
+			}
+
+			encrypted, err := newEncKey.Encrypt(decrypted)
+			if err != nil {
+				return err
+			}
+
+			bucket.Put(k, encrypted)
+
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		// Finally, we can update our database to store the new
+		// encryption key.
+		err = bucket.Put(encryptedKeyID, newEncKey.Marshal())
+		if err != nil {
+			return err
+		}
+
```

Comment:
if the previous `encKey` is not-nil, we should zero it immediately before overwriting it with the new one

Reply:
Fixed.